File: 0146-lru-cache/0146-lru-cache.cpp
struct Node{
    int val;
    int key;
    Node* next;
    Node* prev;
    Node(): val(0), next(nullptr), prev(nullptr){}
    Node(int x): val(x), next(nullptr), prev(nullptr){}
    Node(int x, Node* next, Node* prev): val(x), next(next), prev(prev){}
};

class LRUCache {
private:
    int sz;
    int cap;
    Node* head;
    Node* tail;
    unordered_map<int, Node*> map;
public:
    LRUCache(int capacity) {
        cap = capacity;
        head = new Node();
        tail = new Node();
        head -> next = tail;
        tail -> prev = head;
        sz = 0;
    }
    
    int get(int key) {
        if(map.find(key) != map.end()){
            map[key]->prev->next = map[key] -> next;
            map[key]->next->prev = map[key] -> prev;
            map[key]->prev = head;
            map[key]->next = head -> next;
            head->next->prev = map[key];
            head->next = map[key];
            return map[key]->val;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if(map.find(key) != map.end()){
            map[key]->prev->next = map[key] -> next;
            map[key]->next->prev = map[key] -> prev;
            map[key]->prev = head;
            map[key]->next = head -> next;
            head->next->prev = map[key];
            head->next = map[key];
            map[key] -> val = value;
        }
        else{
            sz++;
            Node* n = new Node(value, head->next, head);
            n -> key = key;
            head->next->prev = n;
            head->next = n;
            map[key] = n;
            if(sz > cap){
                map.erase(tail -> prev -> key);
                tail -> prev -> prev -> next = tail;
                tail -> prev = tail -> prev -> prev;
                sz--;
            }
        }
        
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, prev: 'Node' = None, next: 'Node' = None, key: int = 0, val: int = -1):
        self.prev = prev
        self.next = next
        self.key = key
        self.val = val

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.size = 0
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.map = dict()

    def get(self, key: int) -> int:
        if key in self.map:
            self.__prioritize(key)
            return self.map[key].val
        return -1
        
    def put(self, key: int, value: int) -> None:
        if key in self.map:
            self.__prioritize(key)
            self.map[key].val = value
        else:
            self.size += 1
            if self.size > self.capacity:
                toDelete = self.tail.prev
                self.tail.prev = toDelete.prev
                self.tail.prev.next = self.tail
                del self.map[toDelete.key]

            self.map[key] = Node(key = key, val = value)
            self.__prioritize(key)

    def __prioritize(self, key: int) -> None:
        # print(f"prioritize called on key {key}")
        curr = self.map[key]
        if curr.prev and curr.next:
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
        curr.next = self.head.next
        curr.prev = self.head
        self.head.next.prev = curr
        self.head.next=curr
        
        # tmp = self.head
        # print_list = []
        # while tmp:
        #     print_list.append(str(tmp.val))
        #     tmp = tmp.next
        # out = '->'.join(print_list)
        # print(f'new priorities: {out}')

